src/Analyzer.Core.BuiltInRuleTests/TestRunner.cs

```diff
@@ -37,60 +37,46 @@ public void TestRule(TestConfiguration ruleExpectations)
+            var thisRuleEvaluations = evaluations.Where(e => e.RuleId.Equals(ruleExpectations.TestName, StringComparison.OrdinalIgnoreCase)).ToList();
 
-            // If there are no expected failures, the evaluation should have passed
-            Assert.AreEqual(ruleExpectations.ReportedFailures.Length == 0, thisRuleEvaluation.Passed);
+            // Get all lines reported as failed
+            var failingLines = thisRuleEvaluations
+                .Where(e => !e.Passed)
+                .SelectMany(e => GetAllFailedLines(e))
+                .ToHashSet();
 
-            if (!thisRuleEvaluation.Passed)
-            {
-                // Get all lines reported as failed
-                var failingLines = thisRuleEvaluation.Evaluations
-                    .Where(e => !e.Passed)
-                    .Select(e => GetAllResults(e))
-                    .SelectMany(rs => rs.Where(r => !r.Passed)
-                                    .Select(r => r.LineNumber)
-                                    .ToList())
-                    .ToHashSet();
-
-                failingLines.UnionWith(thisRuleEvaluation.Results
-                    .Where(r => !r.Passed)
-                    .Select(r => r.LineNumber)
-                    .ToHashSet());
-
-                // Verify all expected lines are reported
-                var expectedLines = ruleExpectations.ReportedFailures.Select(failure => failure.LineNumber).ToHashSet();
-                Assert.IsTrue(failingLines.SetEquals(expectedLines),
-                    "Expected failing lines do not match actual failed lines.  " +
-                    $"Expected: [{string.Join(",", expectedLines)}]  Actual: [{string.Join(",", failingLines)}]");
-            }
+            // Verify all expected lines are reported
+            var expectedLines = ruleExpectations.ReportedFailures.Select(failure => failure.LineNumber).ToHashSet();
+            Assert.IsTrue(failingLines.SetEquals(expectedLines),
+                "Expected failing lines do not match actual failed lines.  " +
+                $"Expected: [{string.Join(",", expectedLines)}]  Actual: [{string.Join(",", failingLines)}]");
         }
 
         /// <summary>
         /// Recursively get all <see cref="IResult"/> objects inside an <see cref="IEvaluation"/>.
         /// </summary>
         /// <param name="evaluation">The <see cref="IEvaluation"/> to get the <see cref="IResult"/> from.</param>
         /// <returns>All <see cref="IResult"/>s in the <see cref="IEvaluation"/>.</returns>
-        private IEnumerable<IResult> GetAllResults(IEvaluation evaluation)
+        private IEnumerable<int> GetAllFailedLines(IEvaluation evaluation)
         {
-            foreach (var result in evaluation.Results)
+            if (!evaluation.Result?.Passed ?? false)
             {
-                yield return result;
+                yield return evaluation.Result.LineNumber;
```

Comment:
I'm not quite certain what you're getting at with the question, but if you're asking if just this line can be changed to a regular return since it's not in a loop, the answer is no because if `yield return` is used in a function, every return in the function has to be a `yield return`.

If that's not what you're asking, can you clarify?

Reply:
Ah, thanks. I may not have looked closely at the call stack and saw only the `foreach` loop change